cli: support option to signal a job a configurable time before expiration

grondo opened the pull request:
This PR adds support for a new `--signal=SIG@TIME` option to cli submission commands to optionally send a signal to a job a configurable time before the job time limit.

This implementation uses a job shell option  and plugin to implement the timer which generates the signal. In retrospect, it may have been better to implement this feature in the job-exec module, since that module delivers the `SIGALRM` signal when a job reaches its expiration, and that module currently adjusts the expiration by the delay between the creation of _R_ (the starttime) and the actual time at which job shells started.

Another potential benefit to moving this signal delivery to the exec system is that we then have only one place to update the job expiration when that feature is supported.

It would be fairly simple to make the change to this PR if that is desired. However, posting this here as is because it currently works, and I think up to a few seconds early delivery of the configured signal is probably not a showstopper here. Therefore, it might be better to have something working than perfect.

Fixes #5104

chu11 commented on t/t2618-job-shell-signal.t:
```diff
@@ -0,0 +1,77 @@
+test_expect_success 'shell: signal option works with alternate signal' '
+	test_fails_by_signal 12 flux run -t 5s --signal=USR2@4.5s sleep 15
+'
+test_expect_success 'shell: signal shell option can be an integer' '
+	test_fails_by_signal 10 flux run -t 5s -o signal -o verbose sleep 10 \
+		2>sig.log &&
+	grep "will expire in 60.0s" sig.log
+'
```
i think this test is just testing the defaults, not really testing if integer input acceptable?   Perhaps should have one of each test.

Also, perhaps a test for when `-t` not specified, so signal option shouldn't do anything.

grondo replied:
> i think this test is just testing the defaults, not really testing if integer input acceptable? Perhaps should have one of each test.

No, if `-o option` is used on the command like the default is to set `attributes.system.shell.options.option = 1`:

```console
$ src/cmd/flux run -o signal --dry-run hostname | jq .attributes.system.shell.options.signal
1
```

> Also, perhaps a test for when -t not specified, so signal option shouldn't do anything.

Good idea! I'll add that.

grondo commented on src/shell/signal.c:
```diff
@@ -0,0 +1,257 @@
+     * event is posted, (after all job shells have started). This shell
+     * plugin does *not* do a similar adjustment, so the signal may be
+     * delivered early. However, this difference should be no more than
+     * a few seconds, which should be critical in real-world scenarios.
```
"should not be no more than a few seconds" (if I'm looking at the same text) would be a double negative (and would be the opposite of what I meant here). However, maybe "this difference should be at most a few seconds" would be more clear?

grondo replied:
Oh, I see I read the first "should be" and not the second :facepalm: Will fix!

grondo commented on src/bindings/python/flux/cli/base.py:
```diff
@@ -20,6 +20,7 @@
 import os
 import re
 import resource
+import signal
 import sys
 from collections import ChainMap
 from itertools import chain
```
I started with SIG@TIMELEFT  but it seemed overly verbose. The help description says "TIME before job expiration" and I figured that was good enough.

grondo commented:
I addressed @chu11's comments and force-pushed the result.

chu11 reviewed (approved):
LGTM, although don't have much of an opinion on if it should be moved into `job-exec`.  Perhaps someone else has a strong opinion on it?

garlick commented:
> Perhaps someone else has a strong opinion on it?

I think this is probably fine as is for now or maybe forever. I guess if it becomes a problem we know what to do.  By then maybe the planned exec system upheaval will be complete.

I'll add my review shortly, but at first look it seems great.

garlick commented:
Your recent work (#5202) makes it so that a signal delivered to a batch instance gets propagated to all jobs running in it, so `--signal` is expected to be specified in the batch script or batch submission line correct?  (As opposed to the `flux run` commands or whatever within the batch script?)

I can't remember if we assign a duration to jobs when the enclosing instance has a known time limit?

garlick commented on src/common/libutil/Makefile.am:
```diff
@@ -100,7 +100,9 @@ libutil_la_SOURCES = \
 	strstrip.c \
 	strstrip.h \
```
Commit message for "add signal name/number helper routines" has the return value as `int` for `signutil_signame()`.  Should be `const char *`

garlick commented on src/common/libutil/sigutil.c:
```diff
@@ -0,0 +1,127 @@
+
+#include "ccan/str/str.h"
+#include "ccan/array_size/array_size.h"
+
```
sigutil.h should be included to ensure prototypes match definitions.

grondo replied:
Oh, good catch.

grondo commented:
> Your recent work (https://github.com/flux-framework/flux-core/pull/5202) makes it so that a signal delivered to a batch instance gets propagated to all jobs running in it, so --signal is expected to be specified in the batch script or batch submission line correct? (As opposed to the flux run commands or whatever within the batch script?)

It can be either I think, though I believe the more common use case is with `flux batch` (?) #5202  was therefore a prerequisite for this work.

The scheduler does propagate any time limit of the current instance to the job in question when creating R. This can be seen with this test:

```console
$ flux alloc -n1 -t 15s flux run flux job timeleft
12
[detached: session exiting]
```

Though that doesn't require that `--signal` be used on the `flux run` line -- it would be optional in all cases.

garlick commented on src/shell/signal.c:
```diff
@@ -0,0 +1,257 @@
+    flux_watcher_destroy (sig->watcher);
+    sig->watcher = NULL;
+
+    if (expiration == 0.)
+        return 0;
```
If the user specified `--signal` on a job with no expiration, should something be logged like
> job has no expiration time, SIG will not be sent

grondo replied:
Great idea, I'll add that.

grondo commented:
Oh, are you saying that the option is possibly unnecessary for `flux run` and `flux submit`? I guess for most cases that could be true, but since the `-t, --time-limit` option is offered on those commands, I reasoned that the `--signal` option should be present as well.

garlick reviewed (approved):
LGTM!   Just a couple minor nits.

garlick commented:
> Oh, are you saying that the option is possibly unnecessary for flux run and flux submit

No no, I was just not remembering that the scheduler sets the expiration when none is specified, and it worried me that people might think they need to specify it on the run/submit line and it would be silently ignored.  I was just confused, never mind!

grondo commented:
> might think they need to specify it on the run/submit line and it would be silently ignored.

I like your idea of issuing a warning in this case. I've added that and fixed your other comments and force-pushed the result.

Thanks @chu11 and @garlick!

grondo commented:
Ok, unless I hear something I'll set MWP soon.
